spec/app/ics-721-nft-transfer/README.md

```diff
@@ -35,29 +35,35 @@ The IBC handler interface & IBC routing module interface are as defined in [ICS
 
 ```typescript
 interface NonFungibleTokenPacketData {
   classId: string
   classUri: string
+  classData: byte[]
   tokenIds: string[]
   tokenUris: string[]
   tokenData: byte[][]
   sender: string
   receiver: string
+  memo: string
 }
 ```
 
 `classId` uniquely identifies the class/collection which the tokens being transferred belong to in the sending chain. In the case of an ERC-1155 compliant smart contract, for example, this could be a string representation of the top 128 bits of the token ID.
 
-`classUri` is optional, but will be extremely beneficial for cross-chain interoperability with NFT marketplaces like OpenSea, where [class/collection metadata](https://docs.opensea.io/docs/contract-level-metadata) can be added for better user experience.
+`classUri` is an optional field which, if present, contains off-chain [class metadata](https://docs.opensea.io/docs/contract-level-metadata) that could be extremely beneficial for cross-chain interoperability with NFT marketplaces like OpenSea.
+
+`classData`is an optional field which, if present, contains opaque on-chain class metadata such as royalty related parameters.
 
 `tokenIds` uniquely identifies some tokens of the given class that are being transferred. In the case of an ERC-1155 compliant smart contract, for example, a `tokenId` could be a string representation of the bottom 128 bits of the token ID.
 
-Each `tokenId` has a corresponding entry in `tokenUris` which, if present, refers to an off-chain resource that is typically an immutable JSON file containing the token's metadata.
+Each `tokenId` has a corresponding entry in `tokenUris` which, if not empty, refers to an off-chain resource that is typically an immutable JSON file containing the token's metadata.
+
+Each `tokenId` has another corresponding entry in `tokenData` which, if not null, contains opaque on-chain application data associated with the token.
 
-Each `tokenId` has another corresponding entry in `tokenData` which, if present, contains some opaque application data associated with the token (e.g., royalty parameters).
+The `memo` field is not used within the token transfer, however, it may be used either for external off-chain users (i.e. exchanges) or for middleware wrapping transfer that can parse and execute custom logic on the basis of the passed in memo. If the memo is intended to be parsed and interpreted by higher-level middleware, then these middlewares are advised to namespace their additions to the memo string so that they do not overwrite each other. Chains should ensure that there is some length limit on the entire packet data to ensure that the packet does not become a DOS vector. However, these do not need to be protocol-defined limits. If the receiver cannot accept a packet because of length limitations, this will lead to a timeout on the sender side.
```

Comment:
```
If the receiver cannot accept a packet because of length limitations, this will lead to a timeout on the sender side.
```
If there is an error in the destination chain `ReceivePacket`, the wrong Ack should be returned instead of timeout

Reply:
might be nice to specify here as well that `memo` is associated with the transfer, not the token. something like:

> the `memo` field is associated with the token transfer, not the token.

Reply:
for CosmWasm contracts, we actually can't do this because too long of a memo field could cause a out-of-gas error during deserialization of the packet, before we can run any custom logic. for example:

```
let packet = from_binary(packet_data);
if packet.memo.is_too_long() {
  ACK_FAIL
}
```

could run out of gas during the `from_binary` step. requiring that we return an ACK-FAIL then makes it almost impossible for this to be correctly implemented in CW. the spec being aspirational is fine (we don't necessarily have to pander to the limitations of every execution environment) though. is there a specific advantage of a ACK-FAIL vs a timeout here? ex: does the SDK ics721 implementation handle those differently?

Reply:
I thought about it, if this verification is verified before the transaction is executed, it will indeed not return Fail Ack, thank you for correcting

Reply:
So we'll leave the length limit check of `memo` out of the ics721 protocol.